browser/ui/views/brave_actions/brave_actions_container.h

```diff
@@ -0,0 +1,105 @@
+  public:
+    BraveActionsContainer(Browser* browser, Profile* profile);
+    ~BraveActionsContainer() override;
+    void Init();
+    void Update();
+    void SetShouldHide(bool should_hide);
+
+    // ToolbarActionView::Delegate
+    content::WebContents* GetCurrentWebContents() override;
+    bool ShownInsideMenu() const override;
+    // Notifies that a drag completed.
+    void OnToolbarActionViewDragDone() override;
+    // Returns the view of the toolbar actions overflow menu to use as a
+    // reference point for a popup when this view isn't visible.
+    views::MenuButton* GetOverflowReferenceView() override;
+    // Returns the preferred size of the ToolbarActionView.
+    gfx::Size GetToolbarActionSize() override;
+    // Overridden from views::DragController (required by ToolbarActionView::Delegate):
+    void WriteDragDataForView(View* sender,
+                              const gfx::Point& press_pt,
+                              ui::OSExchangeData* data) override;
+    int GetDragOperationsForView(View* sender, const gfx::Point& p) override;
+    bool CanStartDragForView(View* sender,
+                            const gfx::Point& press_pt,
+                            const gfx::Point& p) override;
+
+    // ExtensionRegistryObserver:
+    void OnExtensionLoaded(content::BrowserContext* browser_context,
+                          const extensions::Extension* extension) override;
+    void OnExtensionUnloaded(content::BrowserContext* browser_context,
+                            const extensions::Extension* extension,
+                            extensions::UnloadedExtensionReason reason) override;
+
+    // ExtensionActionAPI::Observer
+    // Called when there is a change to the given |extension_action|.
+    // |web_contents| is the web contents that was affected, and
+    // |browser_context| is the associated BrowserContext. (The latter is
+    // included because ExtensionActionAPI is shared between normal and
+    // incognito contexts, so |browser_context| may not equal
+    // |browser_context_|.)
+    void OnExtensionActionUpdated(
+        ExtensionAction* extension_action,
+        content::WebContents* web_contents,
+        content::BrowserContext* browser_context) override;
+
+  private:
+    void AddShields(const extensions::Extension* brave_extension);
+    void RemoveShields();
+
+    bool should_hide_ = false;
+
+    ToolbarActionView* shields_button_view_ = nullptr;
+    // The Browser this LocationBarView is in.  Note that at least
+    // chromeos::SimpleWebViewDialog uses a LocationBarView outside any browser
+    // window, so this may be NULL.
+    Browser* const browser_;
+
+    void OnExtensionSystemReady();
+
+    extensions::ExtensionActionAPI* extension_action_api_;
+    extensions::ExtensionRegistry* extension_registry_;
+    extensions::ExtensionActionManager* extension_action_manager_;
+    ToolbarActionViewController* shields_view_controller_ = nullptr;
```

Comment:
same as `shields_button_view_`.

Reply:
I think we should consider using smart pointer to |shields_view_controller_|.
Or if we can guarantee `RemoveShields()` is always called before `BraveActionsContainer` is destroyed,
just adding DCHECK(!shields_view_controller_) in dtor would be sufficient.